### ISSUE TITLE
No error when resource not used.

### ISSUE THREAD

**Issue by pcostell** [google/google-api-nodejs-client#448]
In particular on the Datastore v1beta2 API, something like:

```js
datastore.datasets.lookup({
  datasetId: 'my-dataset-id',
  keys: [...],
}, function(err, response) {
  ...
});
```

Looks correct, but returns an empty response. However, this call is not actually correct since keys are supposed to be provided as part of the payload. Instead the bad API call gets converted into a request at: `/datastore/v1beta2/datasets/pcostello-cloud/lookup?keys=` with an empty payload.

**Comment by jmdobry** [comment]
I'm guessing since this particular endpoint can take multiple keys, passing 0 keys still works, and so it basically just returns 200 OK. It's up the API if wants to change that behavior, but they probably won't since it would be a breaking change.

This library is generated, and has no extra logic for various APIs—it's just a wrapper—so it's at the mercy of what the APIs support.

If you're looking for a more idiomatic way to interact with Cloud Datastore, I recommend you checkout [gcloud-node](https://googlecloudplatform.github.io/gcloud-node/#/docs/v0.28.0/datastore), the official hand-written Node.js client for Google Cloud Platform APIs. It has good support for Cloud Datastore.

**Comment by jmdobry** [comment]
Closing as this does not appear to be an issue. If you are convinced that this is really a bug, please feel free to re-open the issue and add more information such as:

* Expected behavior
* Actual behavior
* Code samples
* Version numbers
* References docs or other links

Or, consider opening a Pull Request.

__Otherwise, support questions are handled on [Stack Overflow][stackoverflow].__ [![Ask a question on Stackoverflow][overflowimg]][stackoverflow]

[overflowimg]: https://googledrive.com/host/0ByfSjdPVs9MZbkhjeUhMYzRTeEE/stackoveflow-tag.png
[stackoverflow]: http://stackoverflow.com/questions/tagged/google-api-nodejs-client

**Comment by pcostell** [comment]
This is definitely still an issue, but I may be able to clarify it more:

the client doesn't restrict path arguments to those defined in the discovery document.

Looking at Datastore as an example, you can see the discovery document listed [here](https://www.googleapis.com/discovery/v1/apis/datastore/v1beta2/rest). Looking at the snippet for Lookup:
```
"lookup": {
     "id": "datastore.datasets.lookup",
     "path": "{datasetId}/lookup",
     "httpMethod": "POST",
     "description": "Look up some entities by key.",
     "parameters": {
      "datasetId": {
       "type": "string",
       "description": "Identifies the dataset.",
       "required": true,
       "location": "path"
      }
     },
     "parameterOrder": [
      "datasetId"
     ],
     "request": {
      "$ref": "LookupRequest"
     },
     "response": {
      "$ref": "LookupResponse"
     },
     "scopes": [
      "https://www.googleapis.com/auth/cloud-platform",
      "https://www.googleapis.com/auth/datastore",
      "https://www.googleapis.com/auth/userinfo.email"
     ]
    },
```

The document lists a single parameter "datasetId". However, if a user forgets to specify the json request data in a resource element, this data will get put into the query string instead. At some point it looks like this gets dropped (either in the clients or at the apiary level). In this case it seems like we should get a nice error message.

This is particularly error prone for Datastore's Lookup RPC because it is valid to send it no keys. Datastore returns an RPC with a list of found entities and a list of missing keys. So if someone remembers to check, they'll see that missing is also empty and that something else must be going wrong (but there is no way to know beyond code inspection or adding logging in the googleapis client itself and noticing that keys get set on the querystring and not the request).

**Comment by pcostell** [comment]
Note -- I don't seem to have reopen permissions, but I do think this is still an issue and is not WAI.

**Comment by jmdobry** [comment]
Okay I see. The issue is that this library is not proactive in protecting the user from mistakes—it doesn't validate the user's input according to the schema provided by the discovery doc, it just sends whatever the user provides.

I've only just started working on google-api-nodejs-client, but I'm thinking it needs some upgrades. One upgrade would be to validate the user's input, such as verifying required path parameters, optional querystring parameters, and request body data.

I've opened #528 to address this.
